fem/kernels/kGeometry.cpp

```diff
@@ -0,0 +1,375 @@
+// Copyright (c) 2010, Lawrence Livermore National Security, LLC. Produced at
+// the Lawrence Livermore National Laboratory. LLNL-CODE-443211. All Rights
+// reserved. See file COPYRIGHT for details.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability see http://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the GNU Lesser General Public License (as published by the Free
+// Software Foundation) version 2.1 dated February 1999.
+
+#include "../../config/config.hpp"
+
+#include "../../general/okina.hpp"
+#include "../../linalg/kernels/vector.hpp"
+
+#include "kGeometry.hpp"
+#include "../fem.hpp"
+#include "../doftoquad.hpp"
+
+// *****************************************************************************
+MFEM_NAMESPACE
+
+// *****************************************************************************
+template<const int NUM_DOFS,
+         const int NUM_QUAD>
+void kGeom2D(const int,const double*,const double*,double*,double*,double*);
+
+// *****************************************************************************
+template<const int NUM_DOFS,
+         const int NUM_QUAD>
+void kGeom3D(const int,const double*,const double*,double*,double*,double*);
+
+// *****************************************************************************
+typedef void (*fIniGeom)(const int,const double*,const double*,
+                         double*, double*, double*);
+
+// *****************************************************************************
+void kGeom(const int DIM,
+           const int NUM_DOFS,
+           const int NUM_QUAD,
+           const int numElements,
+           const double* dofToQuadD,
+           const double* nodes,
+           double* J,
+           double* invJ,
+           double* detJ)
+{
+   const unsigned int dofs1D = IROOT(DIM,NUM_DOFS);
+   const unsigned int quad1D = IROOT(DIM,NUM_QUAD);
+   const unsigned int id = (DIM<<4)|(dofs1D-2);
+   dbg("DIM=%d",DIM);
+   dbg("quad1D=%d",quad1D);
+   dbg("dofs1D=%d",dofs1D);
+   dbg("id=%d",id);
+   assert(LOG2(DIM)<=4);
+   assert(LOG2(dofs1D-2)<=4);
+   static std::unordered_map<unsigned int, fIniGeom> call =
```

Comment:
You don't really need an unordered_map here.  Since everything is you can just use an array.